File: SuperheroApp/Characters/CharactersContract.swift
//
//  CharactersContract.swift
//  SuperheroApp
//
//  Created by Balazs Varga on 2018. 02. 12..
//  Copyright © 2018. W.UP. All rights reserved.
//

struct CharactersContract {
    
    protocol CharactersPresenter: MvpPresenter {
        
        typealias ViewType = CharactersView
        
        func characters() -> [Character]
        
        func charactersCount() -> Int
        
        func loadCharacters()
        
        func loadMoreCharacters()
    }
    
    protocol CharactersView: MvpView {
        
        func showLoadingIndicator()
        
        func hideLoadingIndicator()
    
        func showMoreLoadingIndicator()
        
        func hideMoreLoadingIndicator()
        
        func refreshCharacters()
        
        func showLoadingCharactersError(message: String)
        
        func showNoCharacters()
    }
}

File: android/src/main/java/digital/wup/android/presentation/ui/characters/CharactersContract.kt
package digital.wup.android.presentation.ui.characters


import android.os.Bundle

import digital.wup.superheroapp.common.characters.domain.model.Character
import digital.wup.superheroapp.common.mvp.MvpPresenter
import digital.wup.superheroapp.common.mvp.MvpView

interface CharactersContract {
    interface CharactersPresenter : MvpPresenter {
        fun takeView(view: CharactersView)

        fun characters(): Array<Character>

        fun charactersCount(): Int

        fun loadCharacters()

        fun loadMoreCharacters()
    }

    interface CharactersView : MvpView {
        fun showLoadingIndicator()

        fun hideLoadingIndicator()

        fun showMoreLoadingIndicator()

        fun hideMoreLoadingIndicator()

        fun showCharacters(characters: Array<Character>)

        fun showLoadingCharactersError()

        fun showNoCharacters()

        fun navigateToDetails(bundle: Bundle)
    }
}
